# Progetto Data Science
## Fase 1 - Discovery

Importiamo le librerie che useremo ed il primo dataset

In [29]:
import pandas as pd
import numpy as np
import matplotlib as plt
df = pd.read_csv('./data/hfi_cc_2019.csv')

Controlliamo alcuni campioni all'interno del dataset, ne verifichiamo dimensioni, colonne ed indici

In [2]:
df.sample(3)

year ISO_code        countries                         region hf_score  \
643   2014      VEN        Venezuela  Latin America & the Caribbean      4.5   
1270  2010      LKA        Sri Lanka                     South Asia     6.39   
602   2014      PNG  Pap. New Guinea                        Oceania     6.61   

     hf_rank hf_quartile pf_rol_procedural pf_rol_civil pf_rol_criminal  ...  \
643      158           4               1.5          3.5             1.6  ...   
1270     113           3               4.1          5.2             6.2  ...   
602       95           3                 -            -               -  ...   

     ef_regulation_business_adm ef_regulation_business_bureaucracy  \
643                         0.5                                  0   
1270                        4.5                                6.1   
602                         3.4                                4.7   

     ef_regulation_business_start ef_regulation_business_bribes  \
643                           5.7                           1.5   
1270                          8.8                             5   
602                           8.1                           3.1   

     ef_regulation_business_licensing ef_regulation_business_compliance  \
643                                 5                               1.1   
1270                              7.3                               7.1   
602                               7.4                               7.7   

     ef_regulation_business ef_regulation ef_score ef_rank  
643                     2.3             3     2.96     159  
1270                    6.5           6.8     6.53      99  
602                     5.7             7     6.33     115  

[3 rows x 120 columns]

In [3]:
df.shape

(1620, 120)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1620 entries, 0 to 1619
Columns: 120 entries, year to ef_rank
dtypes: int64(1), object(119)
memory usage: 1.5+ MB


In [5]:
df.columns

Index(['year', 'ISO_code', 'countries', 'region', 'hf_score', 'hf_rank',
       'hf_quartile', 'pf_rol_procedural', 'pf_rol_civil', 'pf_rol_criminal',
       ...
       'ef_regulation_business_adm', 'ef_regulation_business_bureaucracy',
       'ef_regulation_business_start', 'ef_regulation_business_bribes',
       'ef_regulation_business_licensing', 'ef_regulation_business_compliance',
       'ef_regulation_business', 'ef_regulation', 'ef_score', 'ef_rank'],
      dtype='object', length=120)

In [6]:
df.index

RangeIndex(start=0, stop=1620, step=1)

Verifichiamo la presenza di eventuali valori nulli all'interno del dataset

In [7]:
df.isna().values.any()
#Stesso significato di → df.isnull().values.any()

False

## Esito Fase 1
Possiamo notare che ci sono 1620 righe che dovrebbero rappresentare gli stati di tutto il mondo, nel mondo ci dovrebbero esserci circa meno di 200 stati riconosciuti.
Per cui mi sorge il dubbio come mai ci sono così tanti dati in più?

Osservando le righe del dataset...
La presenza dei NaN è limitata in quanto al posto del valore mancante sono stai usati i - , per questo non si vedono richiamando il metodo per visualizzare i valori NaN.

Notiamo inoltre che vi sono 120 colonne
La maggior parte dei dati si presenta di tipo object, anche se rappresenta valori numerici.
Parecchie variabili da tenere in considerazione.

Per questi motivi bisognerà fare un'accurato data cleaning.

---

Verifichiamo perchè ci sono 1620 righe

In [8]:
df['year'].value_counts(dropna=False)

2017    162
2016    162
2015    162
2014    162
2013    162
2012    162
2011    162
2010    162
2009    162
2008    162
Name: year, dtype: int64

Possiamo ricavare con questo comando tutte le righe che hanno una caratteristica in comune

In [9]:
df.iloc[np.flatnonzero(df['hf_quartile'] == '-')]

year ISO_code    countries                         region hf_score  \
336   2015      BLR      Belarus                 Eastern Europe        -   
394   2015      IRQ         Iraq     Middle East & North Africa        -   
461   2015      SDN        Sudan             Sub-Saharan Africa        -   
498   2014      BLR      Belarus                 Eastern Europe        -   
556   2014      IRQ         Iraq     Middle East & North Africa        -   
...    ...      ...          ...                            ...      ...   
1474  2008      SDN        Sudan             Sub-Saharan Africa        -   
1475  2008      SUR     Suriname  Latin America & the Caribbean        -   
1476  2008      TJK   Tajikistan        Caucasus & Central Asia        -   
1477  2008      TLS  Timor-Leste                     South Asia        -   
1478  2008      YEM  Yemen, Rep.     Middle East & North Africa        -   

     hf_rank hf_quartile pf_rol_procedural pf_rol_civil pf_rol_criminal  ...  \
336        -           -                 -            -               -  ...   
394        -           -                 -            -               -  ...   
461        -           -                 -            -               -  ...   
498        -           -                 -            -               -  ...   
556        -           -                 -            -               -  ...   
...      ...         ...               ...          ...             ...  ...   
1474       -           -                 -            -               -  ...   
1475       -           -                 -            -               -  ...   
1476       -           -                 -            -               -  ...   
1477       -           -                 -            -               -  ...   
1478       -           -                 -            -               -  ...   

     ef_regulation_business_adm ef_regulation_business_bureaucracy  \
336                           -                                  -   
394                           -                                  -   
461                           -                                  -   
498                           -                                  -   
556                           -                                  -   
...                         ...                                ...   
1474                          -                                  -   
1475                          -                                  -   
1476                          -                                  -   
1477                          -                                  -   
1478                          -                                  -   

     ef_regulation_business_start ef_regulation_business_bribes  \
336                             -                             -   
394                             -                             -   
461                             -                             -   
498                             -                             -   
556                             -                             -   
...                           ...                           ...   
1474                            -                             -   
1475                            -                             -   
1476                            -                             -   
1477                            -                             -   
1478                            -                             -   

     ef_regulation_business_licensing ef_regulation_business_compliance  \
336                                 -                                 -   
394                                 -                                 -   
461                                 -                                 -   
498                                 -                                 -   
556                                 -                                 -   
...                               ..

In [10]:
df.iloc[np.flatnonzero(df['ISO_code'] == 'BLR')]

year ISO_code countries          region hf_score hf_rank hf_quartile  \
12    2017      BLR   Belarus  Eastern Europe     6.65      92           3   
174   2016      BLR   Belarus  Eastern Europe     6.16     122           3   
336   2015      BLR   Belarus  Eastern Europe        -       -           -   
498   2014      BLR   Belarus  Eastern Europe        -       -           -   
660   2013      BLR   Belarus  Eastern Europe        -       -           -   
822   2012      BLR   Belarus  Eastern Europe        -       -           -   
984   2011      BLR   Belarus  Eastern Europe        -       -           -   
1146  2010      BLR   Belarus  Eastern Europe        -       -           -   
1308  2009      BLR   Belarus  Eastern Europe        -       -           -   
1458  2008      BLR   Belarus  Eastern Europe        -       -           -   

     pf_rol_procedural pf_rol_civil pf_rol_criminal  ...  \
12                 3.9          5.9               5  ...   
174                3.9          5.9               5  ...   
336                  -            -               -  ...   
498                  -            -               -  ...   
660                  -            -               -  ...   
822                  -            -               -  ...   
984                  -            -               -  ...   
1146                 -            -               -  ...   
1308                 -            -               -  ...   
1458                 -            -               -  ...   

     ef_regulation_business_adm ef_regulation_business_bureaucracy  \
12                            -                                  6   
174                           -                                5.6   
336                           -                                  -   
498                           -                                  -   
660                           -                                  -   
822                           -                                  -   
984                           -                                  -   
1146                          -                                  -   
1308                          -                                  -   
1458                          -                                  -   

     ef_regulation_business_start ef_regulation_business_bribes  \
12                            9.7                             -   
174                           9.8                             -   
336                             -                             -   
498                             -                             -   
660                             -                             -   
822                             -                             -   
984                             -                             -   
1146                            -                             -   
1308                            -                             -   
1458                            -                             -   

     ef_regulation_business_licensing ef_regulation_business_compliance  \
12                                8.3                               7.9   
174                                 9                               7.9   
336                                 -                                 -   
498                                 -                                 -   
660                                 -                                 -   
822                                 -                                 -   
984                                 -                                 -   
1146                                -                                 -   
1308                                -                                 -   
1458                                -                                 -   

     ef_regulation_business ef_regulation ef_score ef_rank  
12                        8           7.5     6.64      99  
174                     8.1  

Possiamo estrapolare i valori contenuti per una caratteristica all'interno del dataset

In [11]:
df.groupby('region').count()

year  ISO_code  countries  hf_score  hf_rank  \
region                                                                        
Caucasus & Central Asia          60        60         60        60       60   
East Asia                        60        60         60        60       60   
Eastern Europe                  220       220        220       220      220   
Latin America & the Caribbean   260       260        260       260      260   
Middle East & North Africa      190       190        190       190      190   
North America                    20        20         20        20       20   
Oceania                          40        40         40        40       40   
South Asia                      170       170        170       170      170   
Sub-Saharan Africa              420       420        420       420      420   
Western Europe                  180       180        180       180      180   

                               hf_quartile  pf_rol_procedural  pf_rol_civil  \
region                                                                        
Caucasus & Central Asia                 60                 60            60   
East Asia                               60                 60            60   
Eastern Europe                         220                220           220   
Latin America & the Caribbean          260                260           260   
Middle East & North Africa             190                190           190   
North America                           20                 20            20   
Oceania                                 40                 40            40   
South Asia                             170                170           170   
Sub-Saharan Africa                     420                420           420   
Western Europe                         180                180           180   

                               pf_rol_criminal  pf_rol  ...  \
region                                                  ...   
Caucasus & Central Asia                     60      60  ...   
East Asia                                   60      60  ...   
Eastern Europe                             220     220  ...   
Latin America & the Caribbean              260     260  ...   
Middle East & North Africa                 190     190  ...   
North America                               20      20  ...   
Oceania                                     40      40  ...   
South Asia                                 170     170  ...   
Sub-Saharan Africa                         420     420  ...   
Western Europe                             180     180  ...   

                               ef_regulation_business_adm  \
region                                                      
Caucasus & Central Asia                                60   
East Asia                                              60   
Eastern Europe                                        220   
Latin America & the Caribbean                         260   
Middle East & North Africa                            190   
North America                                          20   
Oceania                                                40   
South Asia                                            170   
Sub-Saharan Africa                                    420   
Western Europe                                        180   

                               ef_regulation_business_bureaucracy  \
region                                                              
Caucasus & Central Asia                                        60   
East Asia                                                      60   
Eastern Europe                                                220   
Latin America & the Caribbean                                 260   
Middle East & North Africa                                    190   
North America                                                  20   
Oceania                                                        40   
South Asia            

In [12]:
df.pf_rol_civil.describe()

count     1620
unique      62
top          -
freq       635
Name: pf_rol_civil, dtype: object

In [13]:
hfi_2019_cleaned = df.copy()

In [14]:
hfi_2019_cleaned.shape

(1620, 120)

In [15]:
my_array = np.flatnonzero(df['hf_score'] == '-')
#my_array.shape
for i in my_array:
    hfi_2019_cleaned.drop(i, inplace=True)

In [16]:
hfi_2019_cleaned.shape

(1540, 120)

In [17]:
hfi_2019_cleaned.hf_score.describe()

count     1540
unique     406
top       6.75
freq        16
Name: hf_score, dtype: object

# Indaghiamo le regioni del mondo che vi sono nel dataset

In [22]:
hfi_2019_cleaned.region.unique()

array(['Eastern Europe', 'Middle East & North Africa',
       'Sub-Saharan Africa', 'Latin America & the Caribbean',
       'Caucasus & Central Asia', 'Oceania', 'Western Europe',
       'South Asia', 'North America', 'East Asia'], dtype=object)

In [31]:
hfi_2019_cleaned.region.value_counts()

Sub-Saharan Africa               390
Latin America & the Caribbean    258
Eastern Europe                   212
Western Europe                   180
Middle East & North Africa       169
South Asia                       153
East Asia                         60
Caucasus & Central Asia           58
Oceania                           40
North America                     20
Name: region, dtype: int64

In [40]:
my_array_1 = df.region.value_counts()
type(my_array_1)

pandas.core.series.Series

In [42]:
my_list = df.region.value_counts()
my_list.to_numpy()

array([420, 260, 220, 190, 180, 170,  60,  60,  40,  20], dtype=int64)

In [43]:
my_list_1 = hfi_2019_cleaned.region.value_counts()
my_list_1.to_numpy()

array([390, 258, 212, 180, 169, 153,  60,  58,  40,  20], dtype=int64)

In [48]:
for i in len(my_list):
    print(my_list[i]-my_list_1[i])

TypeError: 'int' object is not iterable

In [54]:
hfi_2019_cleaned['hf_score'].describe()

count     1540
unique     406
top       6.75
freq        16
Name: hf_score, dtype: object